src/Raven.Server/Documents/ShardedHandlers/ShardedSmugglerHandler.cs

```diff
@@ -0,0 +1,182 @@
+                catch (Exception e)
+                {
+                    if (Logger.IsOperationsEnabled)
+                        Logger.Operations("Export failed .", e);
+
+                    result.AddError($"Error occurred during export. Exception: {e.Message}");
+                    await WriteResultAsync(context, result, ResponseBodyStream());
+
+                    HttpContext.Abort();
+                }
+            }
+        }
+
+        public async Task<IOperationResult> ExportShardedDatabaseInternalAsync(
+            DatabaseSmugglerOptionsServerSide options,
+            BlittableJsonReaderObject blittableJson,
+            long startDocumentEtag,
+            long startRaftIndex,
+            Action<IOperationProgress> onProgress,
+            JsonOperationContext jsonOperationContext,
+            OperationCancelToken token)
+        {
+            var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            var oldOperateOnType = options.OperateOnTypes;
+
+            var operateOnTypes = options.OperateOnTypes &= ~(DatabaseItemType.DatabaseRecord |
+                                                             DatabaseItemType.Subscriptions |
+                                                             DatabaseItemType.Identities |
+                                                             DatabaseItemType.Indexes |
+                                                             DatabaseItemType.ReplicationHubCertificates);
+
+            blittableJson = CreateNewOptionBlittableJsonReaderObject(blittableJson, jsonOperationContext, nameof(options.OperateOnTypes), operateOnTypes);
+
+            using (var outputStream = GetOutputStream(ResponseBodyStream(), options))
+            using (var exportOutputStream = new GZipStream(outputStream, CompressionMode.Compress))
+            {
+                await exportOutputStream.WriteAsync(Encoding.UTF8.GetBytes($"{{  \"BuildVersion\" : {ServerVersion.Build}"));
+                for (int i = 0; i < ShardedContext.ShardCount; i++)
+                {
+                    if (i == ShardedContext.ShardCount - 1) // Last shard need to bring all server wide information
+                    {
+                        blittableJson = CreateNewOptionBlittableJsonReaderObject(blittableJson, jsonOperationContext, nameof(options.OperateOnTypes), oldOperateOnType);
+                    }
+
+                    var cmd = new ShardedStreamCommand(this, async stream =>
+                    {
+                        using (var gzipStream = new GZipStream(GetInputStream(stream, options), CompressionMode.Decompress))
+                        {
+                            var firstLoop = true;
+                            await gzipStream.CopyToAsync(exportOutputStream);
+                        }
+
+                    }, tcs, blittableJson);
+
+                    await ShardedContext.RequestExecutors[i].ExecuteAsync(cmd, jsonOperationContext);
```

Comment:
What happened if this fail? what if one shard is down?

Imagine we have 50 shards.. a small network glitch will force us to start everything from scratch.
I think we need to send the file once we assembled all its pieces first.. I know it might be tricky though.

@ayende thoughts?

Reply:
If one shard is down, we fail. Note that shards are actually database groups, so you'll need all nodes there to be down.
That is fine and acceptable for us.